### PR TITLE
oneshot optimizers inspired by Calais guys

### PR DESCRIPTION
**Pull request by teytaud**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context / Related issue

Calais guys have done a great work.
This includes their work, as well as related SOTA stuff.

## How Has This Been Tested (if it applies)

Super cool experiments which show that metacalais is great.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

**Comment by teytaud** [comment]
fyi
@fteytaud
@juliendehos
for some reason I can not add you as reviewers (one day I'll understand, anyway, nevermind :-) )

File: nevergrad/optimization/oneshotcalais.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+from typing import Optional, Union
+import numpy as np
+from scipy import stats
+from ..common.typetools import ArrayLike
+from ..instrumentation import Instrumentation
+from . import sequences
+from . import base
+from .oneshot import *
+
+MetaCalais = SamplingSearch(
```
Comment (jrapin) [inline nevergrad/optimization/oneshotcalais.py]: at some point not too far aways, these should be mostly hidden from main users, it makes the list of algorithms very "noisy", why do we need so many? If it's for Xp we could create them on demand

File: nevergrad/optimization/oneshot.py [hunk 2]
```diff
@@ -65,18 +89,21 @@ class RandomSearchMaker(base.ParametrizedFamily):
 
     # pylint: disable=unused-argument
     def __init__(self, *, middle_point: bool = False, stupid: bool = False,
+                 quasi_opposite: str = "none",
```
Comment (jrapin) [inline nevergrad/optimization/oneshot.py]: It's more "standard" Python to have: `quasi_opposite: Optional[str] = None`
also, I am not fond of the name, maybe "opposition_mode" would be better (currently it's like having a boolean named `true_false` :D )

File: nevergrad/optimization/oneshot.py [hunk 3]
```diff
@@ -65,18 +89,21 @@ class RandomSearchMaker(base.ParametrizedFamily):
...
                  cauchy: bool = False, scale: Union[float, str] = 1.) -> None:
         # keep all parameters and set initialize superclass for print
         assert isinstance(scale, (int, float)) or scale == "random"
         self.middle_point = middle_point
```
Comment (jrapin) [inline nevergrad/optimization/oneshot.py]: Add:
`assert quasi_opposite is None or quasi_opposite in ["quasi", "opposite"]`

File: nevergrad/optimization/oneshot.py [hunk 4]
```diff
@@ -166,12 +204,15 @@ class SamplingSearch(base.ParametrizedFamily):
 
     # pylint: disable=unused-argument
     def __init__(self, *, sampler: str = "Halton", scrambled: bool = False, middle_point: bool = False,
-                 cauchy: bool = False, scale: float = 1., rescaled: bool = False) -> None:
+                 quasi_opposite: str = "none",
+                 cauchy: bool = False, supercalais: bool = False, scale: float = 1., rescaled: bool = False) -> None:
```
Comment (jrapin) [inline nevergrad/optimization/oneshot.py]: is there no other better name that `supercalais`? :s 
+ update docstring

File: nevergrad/optimization/oneshot.py (hunk 2 above)
Comment (teytaud) [reply nevergrad/optimization/oneshot.py]: quasiopposite is standard terminology --- I follow the rest of your recommendation.

File: nevergrad/optimization/oneshot.py (hunk 3 above)
Comment (teytaud) [reply nevergrad/optimization/oneshot.py]: Done.

File: nevergrad/optimization/oneshot.py (hunk 4 above)
Comment (teytaud) [reply nevergrad/optimization/oneshot.py]: Done, as well as other recommendations from you.

File: nevergrad/optimization/oneshot.py (hunk 2 above)
Comment (jrapin) [reply nevergrad/optimization/oneshot.py]: Oh I add not seen you answered this, sorry. Is it standard to define one of the mode, or all the modes? It sounds like "one of them" (the "quasi" mode), that's why I was in favor of changing, I can change back, sorry :s

**Comment by jrapin** [comment]
> for some reason I can not add you as reviewers (one day I'll understand, anyway, nevermind :-) )
I think only people in facebookresearch org can be reviewers, I had the same issue with Daniel before he joined it

**Pull request merged**